rclcpp/src/rclcpp/executors/static_executor_waitable.cpp

```diff
@@ -0,0 +1,264 @@
+            exec_list_.add_waitable(waitable);
+          }
+            return false;
+        });
+    }
+  }
+
+  // Add the executor's waitable to the executable list
+  exec_list_.add_waitable(this_shared_ptr_);
+}
+
+void
+StaticExecutorWaitable::prepare_wait_set()
+{
+  // clear wait set
+  if (rcl_wait_set_clear(p_wait_set_) != RCL_RET_OK) {
+    throw std::runtime_error("Couldn't clear wait set");
+  }
+
+  // The size of waitables are accounted for in size of the other entities
+  rcl_ret_t ret = rcl_wait_set_resize(
+    p_wait_set_, memory_strategy_->number_of_ready_subscriptions(),
+    memory_strategy_->number_of_guard_conditions(), memory_strategy_->number_of_ready_timers(),
+    memory_strategy_->number_of_ready_clients(), memory_strategy_->number_of_ready_services(),
+    memory_strategy_->number_of_ready_events());
+
+  if (RCL_RET_OK != ret) {
+    throw std::runtime_error(
+            std::string("Couldn't resize the wait set : ") + rcl_get_error_string().str);
+  }
+}
+
+void
+StaticExecutorWaitable::refresh_wait_set(std::chrono::nanoseconds timeout)
+{
+  // clear wait set (memeset to '0' all wait_set_ entities
+  // but keeps the wait_set_ number of entities)
+  if (rcl_wait_set_clear(p_wait_set_) != RCL_RET_OK) {
+    throw std::runtime_error("Couldn't clear wait set");
+  }
+
+  if (!memory_strategy_->add_handles_to_wait_set(p_wait_set_)) {
+    throw std::runtime_error("Couldn't fill wait set");
+  }
+
+  rcl_ret_t status =
+    rcl_wait(p_wait_set_, std::chrono::duration_cast<std::chrono::nanoseconds>(timeout).count());
+
+  if (status == RCL_RET_WAIT_SET_EMPTY) {
+    RCUTILS_LOG_WARN_NAMED(
+      "rclcpp",
+      "empty wait set received in rcl_wait(). This should never happen.");
+  }
+  else if (status != RCL_RET_OK && status != RCL_RET_TIMEOUT) {
+    using rclcpp::exceptions::throw_from_rcl_error;
+    throw_from_rcl_error(status, "rcl_wait() failed");
+  }
+}
+
+bool
+StaticExecutorWaitable::add_to_wait_set(rcl_wait_set_t * wait_set)
+{
+  // Add waitable guard conditions (one for each registered node) into the wait set.
+  for (const auto & gc : guard_conditions_) {
+    if (!gc) {
+      throw std::runtime_error("Failed to get waitable guard condition");
+    }
+    rcl_ret_t ret = rcl_wait_set_add_guard_condition(wait_set, gc, NULL);
+
+    if (ret != RCL_RET_OK) {
+      throw std::runtime_error("Executor waitable: couldn't add guard condition to wait set");
+    }
+  }
+  return true;
+}
+
+size_t StaticExecutorWaitable::get_number_of_ready_guard_conditions()
+{
+  return guard_conditions_.size();
+}
+
+void
+StaticExecutorWaitable::add_node_and_guard_condition(
+  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_ptr,
+  rcl_guard_condition_t * node_guard_condition)
+{
+  weak_nodes_.push_back(node_ptr);
+  guard_conditions_.push_back(node_guard_condition);
+}
+
+void
+StaticExecutorWaitable::remove_node_and_guard_condition(
+  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_ptr)
```

Comment:
this looks the same as the `remove_node` function in `executor.cpp` .
if that's so no need to implement it